sdk/nodejs/cmd/run/run.ts

```diff
@@ -255,8 +255,27 @@ export async function run(
 
     span.setAttribute("typescript-enabled", typeScript);
     if (typeScript) {
-        const transpileOnly = (process.env["PULUMI_NODEJS_TRANSPILE_ONLY"] ?? "false") === "true";
         const compilerOptions = tsutils.loadTypeScriptCompilerOptions(tsConfigPath);
+
+        // tanspileOnly controls wether ts-node should do type checking or not.
+        // Users might have a separate build step that runs tsc for type
+        // checking, and don't want to pay the performance cost of type checking
+        // twice. This also enables using swc, which doesn't support type
+        // checking, with ts-node.
+        //
+        // If the `PULUMI_NODEJS_TRANSPILE_ONLY` env variable is set, we use
+        // that to determine the value of `transpileOnly.` Otherwise we use the
+        // `noCheck` compiler option from the tsconfig. Otherwise we default to
+        // ts-node's default, which is to type check.
+        let transpileOnly = undefined;
+        const transpileOnlyEnv = process.env["PULUMI_NODEJS_TRANSPILE_ONLY"];
+        if (transpileOnlyEnv) {
+            transpileOnly = transpileOnlyEnv === "true";
+        } else {
+            // @ts-ignore
+            transpileOnly = compilerOptions.noCheck;
```

Comment:
What happens if we don't specify `transpileOnly` at all to `tsn.register`?

Reply:
It uses the defaults to false https://github.com/TypeStrong/ts-node/blob/ddb05ef23be92a90c3ecac5a0220435c65ebbd2a/website/docs/options.md?plain=1#L130

By passing `undefined` here, we make it work seamlessly for the swc test case where you just specify the ts-node swc option directly in the tsconfig.json https://github.com/pulumi/pulumi/pull/18067/files#diff-b0329545a30ec3def2f030d437d89013412512243dde59f2ca6d2dcef007e69d. If we defaulted to `false` in our code, then this would require the user to also add `noCheck: true` to their config.